controllers/topology/reconcile_state.go

```diff
@@ -288,44 +325,77 @@ func calculateMachineDeploymentDiff(current, desired map[string]*scope.MachineDe
 	return diff
 }
 
+type unstructuredVersionGetter func(obj *unstructured.Unstructured) (*string, error)
+
+type reconcileReferencedObjectInput struct {
+	cluster       *clusterv1.Cluster
+	current       *unstructured.Unstructured
+	desired       *unstructured.Unstructured
+	versionGetter unstructuredVersionGetter
+	opts          []mergepatch.HelperOption
+}
+
 // reconcileReferencedObject reconciles the desired state of the referenced object.
 // NOTE: After a referenced object is created it is assumed that the reference should
 // never change (only the content of the object can eventually change). Thus, we are checking for strict compatibility.
-func (r *ClusterReconciler) reconcileReferencedObject(ctx context.Context, current, desired *unstructured.Unstructured, opts ...mergepatch.HelperOption) error {
+func (r *ClusterReconciler) reconcileReferencedObject(ctx context.Context, in reconcileReferencedObjectInput) error {
 	log := tlog.LoggerFrom(ctx)
 
 	// If there is no current object, create it.
-	if current == nil {
-		log.Infof("Creating %s", tlog.KObj{Obj: desired})
-		if err := r.Client.Create(ctx, desired.DeepCopy()); err != nil {
-			return errors.Wrapf(err, "failed to create %s", tlog.KObj{Obj: desired})
+	if in.current == nil {
+		log.Infof("Creating %s", tlog.KObj{Obj: in.desired})
+		if err := r.Client.Create(ctx, in.desired.DeepCopy()); err != nil {
+			return errors.Wrapf(err, "failed to create %s", tlog.KObj{Obj: in.desired})
 		}
+		r.recorder.Eventf(in.cluster, corev1.EventTypeNormal, "TopologyCreate", "Created %q", tlog.KObj{Obj: in.desired})
```

Comment:
does it worth having TopologyCreate/TopologyDelete/TopologyUpdate into constants?

Reply:
Done!